osutil/disks/disks.go

```diff
@@ -336,59 +337,88 @@ func (d *disk) FindMatchingPartitionUUID(label string) (string, error) {
-		//         device with major same as disk and minor equal to disk minor + 1
-		// step 3. if we hit a device that does not have a partition attribute, then
-		//         we hit another disk, and shall stop searching
-
-		// note that this code assumes that all contiguous major / minor devices
-		// belong to the same physical device, even with MBR and
-		// logical/extended partition nodes jumping to i.e. /dev/sd*5
-
-		// start with the minor + 1, since the major + minor of the disk we have
-		// itself is not a partition
-		currentMinor := d.minor
-		for {
-			currentMinor++
-			partMajMin := fmt.Sprintf("%d:%d", d.major, currentMinor)
-			props, err := udevProperties(filepath.Join("/dev/block", partMajMin))
-			if err != nil && strings.Contains(err.Error(), "Unknown device") {
-				// the device doesn't exist, we hit the end of the disk
-				break
-			} else if err != nil {
-				// some other error trying to get udev properties, we should fail
-				return "", fmt.Errorf("cannot get udev properties for partition %s: %v", partMajMin, err)
+
+		// step 1. find the devpath for the disk, then glob for matching
+		//         devices using the devname in that sysfs directory
+		// step 2. iterate over all those devices and save all the ones that are
+		//         partitions using the partition sysfs file
+		// step 3. for all partition devices found, query udev to get the fs
+		//         label and partition uuid
+
+		udevProps, err := udevProperties(filepath.Join("/dev/block", d.Dev()))
+		if err != nil {
+			return "", err
+		}
+
+		// get the base device name
+		devName := udevProps["DEVNAME"]
+		if devName == "" {
+			return "", fmt.Errorf("cannot get udev properties for device %s, missing udev property \"DEVNAME\"", d.Dev())
+		}
+		// the DEVNAME as returned by udev includes the /dev/mmcblk0 path, we
+		// just want mmcblk0 for example
+		devName = filepath.Base(devName)
+
+		// get the device path in sysfs
+		devPath := udevProps["DEVPATH"]
+		if devPath == "" {
+			return "", fmt.Errorf("cannot get udev properties for device %s, missing udev property \"DEVPATH\"", d.Dev())
+		}
+
+		// glob for /sys/${devPath}/${devName}*
+		paths, err := filepath.Glob(filepath.Join(dirs.SysfsDir, devPath, devName+"*"))
+		if err != nil {
+			return "", fmt.Errorf("internal error getting udev properties for device %s: %v", err, d.Dev())
+		}
+
+		// Glob does not sort, so sort manually to have consistent tests
+		sort.Strings(paths)
+
+		for _, path := range paths {
+			// check if this device is a partition - note that the mere
+			// existence of this file is sufficient to indicate that it is a
+			// partition, the file is the partition number of the device, it
+			// will be absent for pseudo sub-devices, such as the
+			// /dev/mmcblk0boot0 disk device on the dragonboard which exists
+			// under the /dev/mmcblk0 disk, but is not a partition and is
+			// instead a proper disk
+			_, err := ioutil.ReadFile(filepath.Join(path, "partition"))
```

Comment:
Curious why we not use "osutil.FileExist()" here? Also (nitpick^3) this could be a single line: `if _, err := ioutil.ReadFile(...); err != nil { continue`

Reply:
yes good point it could be done with osutil.FileExist, I think I will do this in a followup if that's alright